posthog/caching/calculate_results.py

```diff
@@ -49,40 +49,62 @@ def calculate_cache_key(target: Union[DashboardTile, Insight]) -> Optional[str]:
     return generate_insight_cache_key(insight, dashboard)
 
 
-def get_cache_type(filter: FilterType) -> CacheType:
-    if filter.insight == INSIGHT_FUNNELS:
-        return CacheType.FUNNEL
-    elif filter.insight == INSIGHT_PATHS:
-        return CacheType.PATHS
-    elif filter.insight == INSIGHT_RETENTION:
-        return CacheType.RETENTION
+def get_cache_type(filter: Optional[FilterType], query: Optional[Dict]) -> CacheType:
+    if not filter and not query:
+        raise Exception("Cannot generate cache type for insight. Must provide a filter or a query")
+
+    if query and query.get("source"):
+        cache_type = query["source"].get("kind", None)
+    elif query and query.get("kind"):
+        cache_type = query["kind"]
+    elif filter and filter.insight == INSIGHT_FUNNELS:
+        cache_type = CacheType.FUNNEL
+    elif filter and filter.insight == INSIGHT_PATHS:
+        cache_type = CacheType.PATHS
+    elif filter and filter.insight == INSIGHT_RETENTION:
+        cache_type = CacheType.RETENTION
     elif (
-        filter.insight == INSIGHT_TRENDS
-        and isinstance(filter, StickinessFilter)
-        and filter.shown_as == TRENDS_STICKINESS
-    ) or filter.insight == INSIGHT_STICKINESS:
-        return CacheType.STICKINESS
+        (
+            filter
+            and filter.insight == INSIGHT_TRENDS
+            and isinstance(filter, StickinessFilter)
+            and filter.shown_as == TRENDS_STICKINESS
+        )
+        or filter
+        and filter.insight == INSIGHT_STICKINESS
+    ):
+        cache_type = CacheType.STICKINESS
     else:
-        return CacheType.TRENDS
+        cache_type = CacheType.TRENDS
+
+    return cache_type
 
 
 def calculate_result_by_insight(
     team: Team, insight: Insight, dashboard: Optional[Dashboard]
-) -> Tuple[str, str, List[Dict[str, Any]]]:
+) -> Tuple[str, str, List | Dict]:
     filter = get_filter(data=insight.dashboard_filters(dashboard), team=team)
     cache_key = generate_insight_cache_key(insight, dashboard)
-    cache_type = get_cache_type(filter)
+    cache_type = get_cache_type(filter, insight.query)
 
     tag_queries(
         team_id=team.pk,
         insight_id=insight.pk,
         cache_type=cache_type,
         cache_key=cache_key,
     )
-    return cache_key, cache_type, calculate_result_by_cache_type(cache_type, filter, team)
 
+    return cache_key, cache_type, calculate_result_by_cache_type(cache_type, filter, insight.query, team)
+
+
+def calculate_result_by_cache_type(
+    cache_type: CacheType, filter: Filter, query: Optional[Dict], team: Team
+) -> Dict | List[Dict]:
```

Comment:
When can you return one vs the other? Makes the api confusing to consume.

Reply:
Yeah, different queries have different shaped results

If we did split them then the union would only need to be on the query handling side... 

I was thinking that we should unify the shape of a query result before we release it maybe (cc @mariusandra @thmsobrmlr )

Reply:
Ah, actually right now only one query is returning a list and it has no other consumers. Perfect!

https://github.com/PostHog/posthog/pull/14643

Reply:
Hmmm but this isn't true for existing insights...

We have two routes here

1) filter based insights -> `List[Dict[str, Any]]` or  `List | List[Dict] | List[List[Dict]]` or `Optional[Any]` depending on where you look
2) query based insights -> `Dict` but hiding that they do have slightly different shapes (see schema.ts)

Reply:
Once we've replaced filters the method would be

```
query = get_query(data=insight.dashboard_query(dashboard), team=team)
    cache_key = generate_insight_cache_key(insight, dashboard)
    cache_type = get_cache_type(query)

    tag_queries(
        team_id=team.pk,
        insight_id=insight.pk,
        cache_type=cache_type,
        cache_key=cache_key,
    )

    return cache_key, cache_type, process_query(team, query, True)
```

Inside `process_query` the Dict becomes a type from `schema.py` but from the point of view of the cache it loads some JSON from an insight and stores some JSON in the cache